apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,333 @@
+
+Regardless, we still wanted to make sure these wouldn't cause issues when the test suite is ran under `pytest-run-parallel`.
+
+#### Solution
+
+Thankfully, we already have the solution, the `thread_unsafe` marker! One of my latest PRs involved marking up all the remaining tests that were thread-unsafe. That PR got nearly 100 comments over the course of 2-3 weeks, and led to me making more PRs when I realized some of these tests actually could be fixed.
+
+One thing that required some interesting solutions was marking up the entire `np.f2py` module as thread-unsafe. Of course you could go in and put a marker on every test, but that's a bit tedious. Instead, I tried putting a `conftest.py` file in the f2py testing folder that would mark the tests as thread-unsafe with `pytest_itemcollected`.
+
+```python
+# numpy/f2py/tests/conftest.py
+@pytest.hookimpl(tryfirst=True)
+def pytest_itemcollected(item):
+  item.add_marker(pytest.mark.thread_unsafe(reason="f2py tests are thread-unsafe"))
+```
+
+This worked, sort of. With `spin` it worked fine, but when installing NumPy locally using an editable install (`pip install -e . --no-build-isolation`), for some reason this new conftest file overrides the base conftest file in `numpy/` that the test suite needs. My mentor came up with the clever solution of just using the base conftest file and checking that the test was in the f2py directory. Fun stuff!
+
+I definitely submitted this PR at a bad time, overlapping with the Quansight retreat, so shout out to Nathan, Lysandros, and [Ralf Gommers](https://github.com/rgommers) for taking a look at it despite that!
+
+### CI Job
+
+Finally, the last step to get the NumPy test suite running under parallel threads was to get `pytest-run-parallel` running in NumPy's CI workflow. Coming into this internship, I had some experience with GitHub Actions, having experimented with it when making my own personal Python library. So, my task was to go into the GitHub Action files and put `pytest-run-parallel` somewhere.
+
+Easier said than done. We want to have a whole new `pytest` run here, which will increase the time NumPy's CI jobs will run. It already takes a while to run, so we needed to be clever about it and maybe find a `pytest` run that isn't really needed anymore.
+
+After some trial and error, Nathan suggested I try putting it under the accelerated macOS runs. These tests are already testing Python 3.14t, and are somewhat redundant. Perfect! After figuring out how to do if statements, I was able to put in my shiny new test run into NumPy.
```

Comment:
It's not an accelerated run, it's that it uses the Apple "Accelerate" linear algebra framework instead of OpenBLAS. But that detail isn't interesting for what you're saying here, so it'd just say "Nathan suggested I try using one of the existing 3.14t CI jobs".